### ISSUE TITLE
Flexible yet secure SSL/TLS version

### ISSUE THREAD

**Issue by janschatz** [eclipse-paho/paho.mqtt.c#545]
Paho doesn't seem to allow me to enable several SSL/TLS versions unless I chose MQTT_SSL_VERSION_DEFAULT. I don't want to use MQTT_SSL_VERSION_DEFAULT, because it enables SSLv2 and SSLv3, which are both not considered secure anymore. However I also don't want to limit my users to only one of TLSv1.0, TLSv1.1 and TLSv1.2. I want to have them all, because all of them are considered secure at the moment.
Do I have any options?

**Comment by tbeu** [comment]
One option for the library would be to disable SSLv2 permanently and only enable SSLv3 conditionally at compile time. This is the way how OpenSSL considers weak security.

**Comment by icraggs** [comment]
We'll have to decide what the right TLS versions are appropriate now.  1.3 should probably be part of the mix too.   I expect testing to be the most time consuming aspect.

**Comment by tbeu** [comment]
#430 deals with TLS v1.3.

**Comment by icraggs** [comment]
Yes, #430 deals with the changes needed for TLS v1.3, but not the testing :-)

**Comment by janschatz** [comment]
I was more asking from a user's point of view. So I understand there currently is no way to achieve this.

How about adding something like `MQTT_SSL_VERSION_TLS_1_0_OR_HIGHER` etc.? Typically I want to have newer and more secure versions automatically (i.e. with new OpenSSL versions) and just disable older versions on purpose. The creation of the context could look like this in `SSLSocket_createContext`:
```
net->ctx = SSL_CTX_new(TLS_client_method());

switch(sslVersion)
{
  ...
  case MQTT_SSL_VERSION_TLS_1_0_OR_HIGHER:
    SSL_CTX_set_options(net->ctx, SSL_OP_NO_SSLv3);
    // Fall through
  case MQTT_SSL_VERSION_TLS_1_1_OR_HIGHER:
    SSL_CTX_set_options(net->ctx, SSL_OP_NO_TLSv1);
  ...
}
```

**Comment by a-andre** [comment]
Support for TLS 1.0 should be dropped, see https://blog.pcisecuritystandards.org/are-you-ready-for-30-june-2018-sayin-goodbye-to-ssl-early-tls

People depending on that version can use an older version of paho.
